exercises/alphametics/src/test/java/AlphameticsTest.java

```diff
@@ -0,0 +1,129 @@
+import org.junit.Ignore;
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.hamcrest.CoreMatchers.*;
+import static org.junit.Assert.*;
+
+public class AlphameticsTest {
+
+    @Test
+    public void testThreeLetters() {
+        LinkedHashMap<Character, Integer> expected = new LinkedHashMap<>();
+        expected.put('I', 1);
+        expected.put('B', 9);
+        expected.put('L', 0);
+
+        assertEquals(expected, new Alphametics("I + BB == ILL").solve());
+    }
+
+    @Ignore("Remove to run test")
+    @Test
+    public void testUniqueValue() {
+        assertNull(new Alphametics("A == B").solve());
```

Comment:
In cases like this I would expect it to throw an exception instead of returning `null`. For error handling it's okay to diverge a bit from the canonical data (I believe they're trying to remove most error cases from the canonical data for this reason) since it can vary a lot from language to language how you'd want to handle invalid input. In my opinion it's better to throw an exception in Java than to return null since that forces to user to deal with the error instead of relying on them checking if the output is null :)

Reply:
For an example where we've used custom exceptions in cases like this see the [bank account tests](https://github.com/exercism/java/blob/master/exercises/bank-account/src/test/java/BankAccountTest.java). You could also use `IllegalArgumentException` if you would prefer not to use a custom one :)

Reply:
Ah I see! ^^ Currently I see 3 options:
1. Throwing a checked Exception e.g. `BankAccountActionInvalidException`.
2. Throwing an unchecked Exception e.g. `IllegalArgumentException`.
3. Returning an `Optional<Map<Character, Integer>>`.

I don't quite like option 2 because the user isn't supposed to catch an unchecked Exception according to the [javadoc](https://docs.oracle.com/javase/tutorial/essential/exceptions/runtime.html).

Comparing between option 1 & 3, I prefer option 3 because not having a result doesn't seem to be that exceptional. Or perhaps, I guess `Optional` does the job well enough, so we don't have to write an additional `Exception` class.

```java
/**
 * Returns the solution to the alphametics puzzle if it exists.
 */
Optional<Map<Character, Integer>> solve() {
    AlphameticsRecursion solver = new AlphameticsRecursion(getDistinctCharacters());
    solver.generate();
    return solver.get();
}
```

Reply:
I'm not sure that would be the best solution. It's an error case, the user (in this case the tests) has provided invalid input, so I would expect it to throw an exception. It would also be more consistent with the other exercises and won't require the person implementing the exercise to learn about optionals. 

We aren't expecting the anything to catch the exception, so I think it's fine to use an unchecked exception. We only want the tests to confirm that an exception has been thrown (by using `ExpectedException`). I normally prefer using a custom exception because I think that's neater and makes it more clear what's going on since you can give it a relevant name :) In that case we would provide the implementation of the exception as starter implementation. Does that make sense? 

Those are just my preferences, feel free to disagree :)

Reply:
I'd agree - in most of the existing test suites that require this kind of thing, it isn't uncommon for us to use `ExpectedException`.

For consistency (and ease of the users) I probably prefer option 2? But I option 1 sounds good to me too!

Reply:
Having thought about it a bit more, I do see your point @Zhiyuan-Amos that this isn't really a very exceptional case. However, I would still argue for using an exception because:
1. As @sjwarner-bp said, it would be more consistent with other exercises.
2. Exceptions are very well implemented in Java in my opinion, they're one of my favourite parts of the language actually :) Whereas optionals, while very useful in other languages, are a bit of a pain in Java. If we changed the solve method to return an optional we would have to change all the other tests to check whether it's present, or live with the compiler warnings telling us we haven't checked whether it's present.

However, since it's not really a case of an illegal argument, and it's something we would want the user to be able to recover from, you're probably right that it should be a checked exception. Let me know what you think @Zhiyuan-Amos :)

Reply:
Yes, I agree that using an `Exception` instead is more consistent with other exercises. ^^ I've updated it accordingly. Thanks :)